### PR TITLE
Fix Document#referrer override

### PR DESCRIPTION
**Pull request by tartpvule**
Fixes point 2 of #17 and harden the hook.

Refactor main logic into a class in engine.js
Rewrite and harden the `Document#referrer` hook
Use `browser.contentScripts.register()` to config the hook at document_start
Raise the minimum requirement to FF59

The rules and logic are run by both the webRequest handler and the hook in the content script.

Limited usage and testing for now.
Feedback, testing, and questions appreciated.

### CODE REVIEW

File: background.js [hunk 1]
```diff
@@ -133,79 +90,66 @@ function modifyReferer(e)
...
-		case "remove":
-			referrer = '';
-			break;
-		default:
-			/* "keep" */
-			referrer = request.referrer;
+async function refreshConfig(change, area)
+{
+	let dirty = false;
+	if (area === "sync") {
```
Comment (rugk) [inline background.js:99]: `"sync"` is a magic constant here, which is… not so good.

File: background.js [hunk 2]
```diff
@@ -133,79 +90,66 @@ function modifyReferer(e)
...
+		if (change.hasOwnProperty("domains")) {
```
Comment (rugk) [inline background.js:100]: `"domains` in change` should do the same (unless one does some crazy prototype stuff) and is the better syntax IMHO.

File: background.js [hunk 3]
```diff
@@ -133,79 +90,66 @@ function modifyReferer(e)
...
-
-/*
- * Escape function from
- * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions
- */
-function escapeRegExp(string)
-{
-	return string.replace(/[.*+?^${}()|\[\]\\]/g, '\\$&');
-	// $& means the whole matched string
+	if (dirty) {
```
Comment (rugk) [inline background.js:114]:
```suggestion
	if (isDirty) {
```

**Review by rugk — COMMENTED** [review]
Just FYI.

File: background.js (hunk 1 above)
Comment (tartpvule) [reply background.js:99]: The old code used "sync" here.

File: background.js (hunk 2 above)
Comment (tartpvule) [reply background.js:100]: The old code did it this way.

File: background.js (hunk 3 above)
Comment (tartpvule) [reply background.js:114]: Let's wait for airtower-luna's opinion on coding styles.

File: manifest.json [hunk 4]
```diff
@@ -23,21 +23,23 @@
...
 	"background": {
-		"scripts": ["background.js"]
+		"scripts": ["engine.js", "background.js"]
 	},
 
 	"content_scripts": [
 		{
 			"matches": ["https://*/*", "http://*/*"],
-			"js": ["inject-referer.js"],
+			"match_about_blank": true,
```
Comment (airtower-luna) [inline manifest.json:32]: Why should the content scripts run in blank frames?

File: background.js (hunk 3 above)
Comment (airtower-luna) [reply background.js:114]: I prefer single-word variable names, as long as they're sufficiently descriptive. `dirty` fits that, `changed` or something likes that would be better.

File: content.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+"use strict";
+
+// [CAVEAT]
+//  When the dynamic content script is run is undefined.
+//  engineConfig may be prefilled,
+//  or engineInstance.setConfig will be called later.
+var engineConfig;
+var engineInstance = new RefererModEngine(engineConfig);
```
Comment (airtower-luna) [inline content.js:25]: The problem is clear from the comment, but it should describe the (clever!) workaround, too. The `[CAVEAT]` on its own sounds to me like things might explode at any time. :wink: 

This also applies to the comment in the `registerContentScript()` function, but it's sufficient to have details in one place and a "see also" in the other.

File: background.js [hunk 6]
```diff
@@ -133,79 +90,68 @@ function modifyReferer(e)
...
- * global variables.
- */
-function refreshConfig(change, area)
-{
-	if (area === "sync" && change.hasOwnProperty("domains"))
-	{
-		domains = change.domains.newValue;
-		createDomainRegex();
+	if (registeredContentScript) {
+		registeredContentScript.unregister();
```
Comment (airtower-luna) [inline background.js:150]: Why is there another `registeredContentScript.unregister();` call here, after the one in line 133?

File: engine.js [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+		}
+	}
+
+	/*
+	* Compute referrer for a given pair of URL and its origin.
+	*/
+	computeReferrer(url, originUrl) {
+		if (url.startsWith("about:") || url.startsWith("data:")) {
+			return originUrl;
+		}
```
Comment (airtower-luna) [inline engine.js:100]: Why is this check needed? The `matches` filter in the manifest should prevent the script from being loaded for any protocol other than HTTP(S). Is there anything that may evade the filter?

File: background.js (hunk 3 above)
Comment (tartpvule) [reply background.js:114]: I think `dirty` as in a "dirty bit".

File: manifest.json (hunk 4 above)
Comment (tartpvule) [reply manifest.json:32]: A page can embed `<iframe id="myiframe" src="about:blank"></iframe>`, then grab a reference to non-hooked `get referrer` and use that on his own document.
This is to prevent that.

File: content.js (hunk 5 above)
Comment (tartpvule) [reply content.js:25]: - Should I change `[CAVEAT]` to `[WORKAROUND]`?
- I personally think `engineConfig may be prefilled, or engineInstance.setConfig will be called later.` is an apt description here. Should I also include a link to Temporal Dead Zone (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let#temporal_dead_zone_tdz)? What do you have in mind for a description here?

File: background.js (hunk 6 above)
Comment (tartpvule) [reply background.js:150]: Hard to explain this, but I will try...

Both `registeredContentScript.unregister` and `browser.contentScripts.register` are asynchronous operations.

The `registerContentScript` is structured like: ...blahblah... | "preemptable" | `await browser.contentScripts.register` | "preemptable" | ...blahblah...
(I am borrowing the "preemptable" term here, it is not really accurate, but should convey the meaning)

What happens is that the first `registeredContentScript.unregister` occurs "concurrently" to `browser.contentScripts.register`.
In the time we `await browser.contentScripts.register(...)`, it is possible that `registerContentScript` is called again, so without an `registeredContentScript.unregister` at the end, we will end up with double registered content scripts, and one of which is orphaned.

Thinking it again, it might be better to `unregister` after `register`, will improve this.

File: engine.js (hunk 7 above)
Comment (tartpvule) [reply engine.js:100]: It is to cover the "embedding iframe" case I mentioned earlier in the comment above.

File: background.js (hunk 6 above)
Comment (tartpvule) [reply background.js:152]: Uploaded new commits

File: background.js (hunk 3 above)
Comment (airtower-luna) [reply background.js:114]: Like for a cache, yes, but in this case unnecessarily generic. Just a style issue though. :slightly_smiling_face:

File: content.js (hunk 5 above)
Comment (airtower-luna) [reply content.js]: The comment should describe that we can't know which script (this file or the dynamically generated one) runs first, so the code has to handle both cases, and what the approach for that is. The reason for that code isn't obvious, so it needs to be explained, including why the value for the `document.referrer` has to be determined dynamically.

I don't really care about those labels, but if you want to leave one `[WORKAROUND]` would be much better.

Comment (tartpvule) [reply content.js]: > The comment should describe that we can't know which script (this file or the dynamically generated one) runs first, so the code has to handle both cases, and what the approach for that is.

`When the dynamic content script is run is undefined.`
Or should it be something else? I don't know what better to write here.

> including why the value for the document.referrer has to be determined dynamically.

Because 1) there is no way to synchronously get settings from the background script and 2) `get referrer` can be called cross-window/document.
Where in the file should I put the explanation?
To me ...
Number 1 looks like it is obvious from the WebEnteision API documentation
Number 2 looks obvious from the DOM spec.

Do you have a suggestion as to what to write here?

File: manifest.json (hunk 4 above)
Comment (airtower-luna) [reply manifest.json:32]: Shouldn't such an iframe see the enclosing document URL as its referrer? Can I check the behavior somewhere, or better reproduce it in a test?

Comment (tartpvule) [reply manifest.json:32]: > Shouldn't such an iframe see the enclosing document URL as its referrer?

Precisely! That check is there so that an iframe see the enclosing document URL as its referrer.

File: content.js (hunk 5 above)
Comment (airtower-luna) [reply content.js]: > `When the dynamic content script is run is undefined.`
> Or should it be something else? I don't know what better to write here.

The order relative to the static script being undefined is key here. So something like: "It is undefined whether this file will run before or after the dynamic content script (see registerContentScript() in background.js)."

As far as I'm concerned that's also why the referrer to return must be determined dynamically, if we could be sure the dynamic script with the config would run first I don't see why we couldn't use a static value.

> 2) `get referrer` can be called cross-window/document.

This is definitely non-obvious to me, and I'm still not sure about the possible impact here. Is that connected to the blank iframe question? Can you point me at the relevant documentation?

File: manifest.json (hunk 4 above)
Comment (airtower-luna) [reply manifest.json:32]: That's a mean trick! I'll have to try that myself, and it needs to be in a test! :sweat_smile:

File: content.js (hunk 5 above)
Comment (tartpvule) [reply content.js]: > "It is undefined whether this file will run before or after the dynamic content script (see registerContentScript() in background.js)."

OK

> I don't see why we couldn't use a static value.

Wrong, do that cross-iframe code I mentioned in the "above" comment on a vanilla Firefox and see.

> This is definitely non-obvious to me, and I'm still not sure about the possible impact here. Is that connected to the blank iframe question? Can you point me at the relevant documentation?

Yes, it is connected.
Inspect `Reflect.getOwnPropertyDescriptor(Document.prototype)` on vanilla Firefox, and you will see that it is a getter, not a static value.

File: background.js (hunk 6 above)
Comment (airtower-luna) [reply background.js:154]: [According to MDN](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/contentScripts/RegisteredContentScript/unregister) unregister is synchronous, so unregistering the old script only after registering the new one should be fine. Might still lead to weird results if the order of the async register operations resolving could end up different from the order of events causing them.

Comment (tartpvule) [reply background.js:154]: > According to MDN

It did not say "synchronous", did it? Or are you assuming that based on the `Return value -> None.`?
Also, what is important here is the fact that `register` is asynchronous.

BTW, do you know that there is a documentation error there?
Check the return value of `unregister()` in real Firefox and you will see that it is a `Promise`!

File: content.js (hunk 5 above)
Comment (tartpvule) [reply content.js]: > Can you point me at the relevant documentation?

https://searchfox.org/mozilla-central/source/dom/webidl/Document.webidl#128

File: manifest.json (hunk 4 above)
Comment (tartpvule) [reply manifest.json:32]: Sorry, edited code.

Comment (airtower-luna) [reply manifest.json:32]: I've pushed an update to the tests that checks this issue, see the [`iframe-test` branch](https://github.com/airtower-luna/referer-mod/tree/iframe-test). If you merge that into your branch we can see if the modification works. Thank you for the explanation and example! :slightly_smiling_face:

File: content.js (hunk 5 above)
Comment (airtower-luna) [reply content.js]: The comment should be just one paragraph, and two labels are confusing, best just remove them and combine the text so the connection between the sentences becomes clear:

"It is undefined whether static content scripts will run before or after dynamic content scripts. If the dynamic script ran first engineConfig is prefilled here, otherwise it will call engineInstance.setConfig later."

File: manifest.json (hunk 4 above)
Comment (tartpvule) [reply manifest.json:32]: Looks like the tests passed

Comment (airtower-luna) [reply manifest.json:32]: All green, nice!

File: background.js (hunk 6 above)
Comment (airtower-luna) [reply background.js:153]: I just checked and you're right. The documentation doesn't explicitly say anything, and I got used to async function being marked as such. Which leaves the question how to avoid double (or worse) script registrations, in Python I'd just wrap the register/unregister part with an [`asyncio.Lock`](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Lock), but I don't see an obvious equivalent in JS.

Comment (tartpvule) [reply background.js:153]: I think that is solved as of https://github.com/airtower-luna/referer-mod/pull/20/commits/386885f5d9f8d3988e72cc9871d5142853fdba43.

File: content.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+	const dummy = {
+		get referrer() {
+			// `this` is an XPCNativeWrapper instance
+
+			// In case someone calls us on some random things
+			if (_toString.call(this) !== "[object HTMLDocument]" ||
+				_toString.call(Reflect.getPrototypeOf(this)) !== "[object HTMLDocument]"
+			) {
+				return _call.call(originalGetter, this);
+			}
```
Comment (airtower-luna) [inline content.js:46]: In what kind of situation could this happen?

File: background.js (hunk 6 above)
Comment (airtower-luna) [reply background.js:153]: The double unregister call looks a bit annoying, but yes, it should work even in case some other async call set `registeredContentScript` meanwhile. :slightly_smiling_face:

File: content.js (hunk 8 above)
Comment (tartpvule) [reply content.js:46]:
```
var foo = Reflect.getOwnPropertyDescriptor(Document.prototype, 'referrer').get.call({});
var bar = Document.prototype.referrer;
```

File: background.js (hunk 6 above)
Comment (tartpvule) [reply background.js:153]: For more details, you might be interested in
https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API/Microtask_guide/In_depth
https://v8.dev/blog/fast-async#tasks-vs.-microtasks

File: content.js (hunk 8 above)
Comment (airtower-luna) [reply content.js:46]: Okay, I can see the issue. So the "damage" is only that someone could detect the extension, but still good to avoid. :sweat_smile:

Why not use
```js
if (!(this instanceof HTMLDocument))
```
for the check, though?

Comment (tartpvule) [reply content.js:46]: Call me paranoid, but I was accounting for the case where `this` is from other window/document, such as `<iframe>`, `dispatchEvent`.
Admittedly, I have not tested everything thoroughly so I err on the side of paranoid.

Comment (tartpvule) [reply content.js:46]: As for why I don't trust `instanceof`? `Object.create(HTMLDocument.prototype) instanceof HTMLDocument`.
While we are protected by `XPCNativeWrapper` and this might not apply here, I think `instanceof` does too many "magic" for my taste.

Comment (airtower-luna) [reply content.js:46]: Okay, I see your point about `instanceof`. :grin: 

While reading up on this I unfortunately found a workaround for the current approach, too: :sweat_smile: 

```js
function C() {}
C.prototype[Symbol.toStringTag] = 'HTMLDocument';
var c = new C();
Object.prototype.toString.call(c);
Object.prototype.toString.call(Reflect.getPrototypeOf(c));
```

If there isn't any reliable way to check the type and we can't (reliably) prevent detection that the `document.referrer` has been modified, I think the simple check

```js
if (Object.prototype.toString.call(this) !== "[object HTMLDocument]")
```

should be good enough to avoid messing anything up accidentally.

File: content.js [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+//  It is undefined whether static content scripts will run before
+//  or after dynamic content scripts. If the dynamic script ran first,
+//  engineConfig is prefilled here, otherwise engineInstance.setConfig()
+//  will be called later.
+var engineConfig;
+var engineInstance = new RefererModEngine(engineConfig);
+
+(function() {
+
+	const originalGetter = Reflect.getOwnPropertyDescriptor(Document.wrappedJSObject.prototype, "referrer").get;
```
Comment (airtower-luna) [inline content.js:30]: I'm wary of calling anything from the wrapped object, just in case it might have been modified already. This script should run first, but after all we only need it because of timing issues. Retrieving the getter from `Document.prototype` works just as well, do you see any reason not do use that?

I hope this is the last question, but we'll see. :sweat_smile:

File: content.js (hunk 8 above)
Comment (tartpvule) [reply content.js:46]: > While reading up on this I unfortunately found a workaround for the current approach, too:
> ...

Got "[object Object]" in content script.
Cannot fool content scripts due to `XPCNativeWrapper`. See https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Sharing_objects_with_page_scripts#xray_vision_in_firefox

Firefox ROCKS! This will fool content scripts in Chromium.

> `if (Object.prototype.toString.call(this) !== "[object HTMLDocument]")`
> should be good enough to avoid messing anything up accidentally.

Won't catch `HTMLDocument.prototype` then.

File: content.js (hunk 9 above)
Comment (tartpvule) [reply content.js:30]: > I'm wary of calling anything from the wrapped object, just in case it might have been modified already.

The assumption is that this content script is run only at document-start, so there will be no problem.
The only way I can see this assumption being broken is when installing the extension for the first time and you have tabs already open.

> This script should run first, but after all we only need it because of timing issues.

Not just because of "timing issues". That is covered by the dynamic content script registered from background.js
The necessity is more because of the fact that `document.referrer` is not a simple static value but a getter and
1) can change on a live document, if you change things in `about:config`!
2) can be called cross-document

> Retrieving the getter from `Document.prototype` works just as well, do you see any reason not do use that?

Hook chaining.

File: content.js (hunk 8 above)
Comment (tartpvule) [reply content.js:46]: You may also be interested in https://developer.mozilla.org/en-US/docs/Mozilla/Gecko/Script_security for more information on Gecko's security designs.

File: content.js (hunk 9 above)
Comment (airtower-luna) [reply content.js:30]: > can change on a live document, if you change things in `about:config`!

Under which circumstances would that be desirable? I want the `document.referrer` to be exactly the same as what was sent in the HTTP `Referer` header.

> Hook chaining.

Please elaborate, that unfortunately doesn't mean anything to me.

File: content.js (hunk 8 above)
Comment (airtower-luna) [reply content.js:46]: I just confirmed that with some modification to the test environment, thank you! I had tested on the debug tools console before, and there the trick works. :sweat_smile:

The "Xray vision" and "Script security" pages are marked as archived on MDN, so I'm not sure how far we can still rely on them.

File: content.js (hunk 9 above)
Comment (tartpvule) [reply content.js:30]: > I want the `document.referrer` to be exactly the same as what was sent in the HTTP `Referer` header.

There are some edge cases that makes that impossible. By Mozilla's design, I might add. But they all hinges on *user/privileged* actions.
Explanation:
What Referrer values are generated are dependent on the values of preferences documented in https://wiki.mozilla.org/Security/Referrer *at the time they are accessed*. This means that, by design, `document.referrer` might not reflect the HTTP Referer header sent, provided that the user or some other privileged entities change `about:config` preferences after the HTTP headers are sent, but before the page script accesses `document.referrer`.
This means that our content script that depends on Firefox-provided `document.referrer` is impacted too.
BUT! As I mentioned, this can only be a result of privileged entities, so I don't think too much about breaking this behavior.
So... there, an obscure edge case.

> Please elaborate, that unfortunately doesn't mean anything to me.

Suppose the user installs two extensions that both wants to hook `document.referrer`, I think the expectation is that the hooks are chained together, not that one hook overrides another. If you use the getter retrieved from `Document.prototype`, you will miss this case, since your hook now goes directly to the *privileged* getter.
In addition, I don't think it is wise to use *privileged* getter for response to potentially hostile code. Less confused deputy problems that way. Safer to use *unprivileged* getters, if only for hygiene.

File: content.js (hunk 8 above)
Comment (tartpvule) [reply content.js:46]: > The "Xray vision" and "Script security" pages are marked as archived on MDN, so I'm not sure how far we can still rely on them.

As far as I can see (and tested), still all accurate. And I don't think Mozilla can easily replace those mechanisms, they are assumptions made implicitly by half the browser code. Mozilla will probably have to rewrite Firefox to invalidate those pages.

File: content.js (hunk 9 above)
Comment (tartpvule) [reply content.js:30]: More on hook chaining...
It's the idea behind Windows' [SetWindowsHookEx](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowshookexw)

Quote:
```
Calling the CallNextHookEx function to chain to the next hook procedure is optional,
but it is highly recommended; otherwise, other applications that have installed hooks
will not receive hook notifications and may behave incorrectly as a result. You should
call CallNextHookEx unless you absolutely need to prevent the notification from
being seen by other applications.
```

Comment (airtower-luna) [reply content.js:30]: > In addition, I don't think it is wise to use privileged getter for response to potentially hostile code. Less confused deputy problems that way. Safer to use unprivileged getters, if only for hygiene.

That depends on in what context the getter runs: I want to be sure that a getter from some other source (worst case a page script) won't run get to run in add-on context. Both https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Sharing_objects_with_page_scripts and https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Language_Bindings/Components.utils.exportFunction seem to *imply* that functions retain their source context when passed around, but they're not very clear on the matter. If the page context getter definitely runs in its unprivileged context even if called from the content script via reflection it's not an issue. :slightly_smiling_face: Do you have any better documentation on the matter?

I see what you mean by "hook chaining" now. Clean interaction with other add-ons is definitely desirable, but only as long as it doesn't cause security issues.

**Comment by airtower-luna** [comment]
I've merged your changes with some modifications to the test to verify that the timing issue that existed previously is gone (see https://github.com/airtower-luna/referer-mod/commit/8aec42514fe68a84d9a30acc87bb1025f9d21598, `dynamic-content-script` branch). I'm happy to say the `document.referrer` modification works reliably now, but the iframe trick works when the test is doing the same navigation twice: https://github.com/airtower-luna/referer-mod/runs/2335313324

Do you have any idea what might be causing that?

File: content.js (hunk 9 above)
Comment (tartpvule) [reply content.js:30]: > Do you have any better documentation on the matter?

Sadly, no. But I can say that the xray environment is a mess in some semi-documented/undocumented way. Not all objects in the content script space is created equal.
A promient example I found is in the interaction between privileged and unprivileged `Reflect.apply` and `Array` instances with functions in them; an unprivileged `Reflect.apply` will not be able to access that unprivileged function even when being used to call on a non-xray-wrapped unprivileged function, but a privileged `Reflect.apply` will be able to.

Reading between the lines https://developer.mozilla.org/en-US/docs/Mozilla/Gecko/Script_security, you will see that JavaScript globals such as `Reflect`, `Object`, etc. are different between each contexts. This means content script's `Object` is not the same as `window.Object` and, more confusingly, `window.Object.create(...bla...)` is not the same as `window.wrappedJSObject.Object.create(...bla...)`; the former is creating an xray-wrapped *privileged* object, but the latter is creating non-xray-wrapped *unprivileged* object.

As far as I can tell, DOM objects are placed inconsistently. For some objects, `window.Foo.wrappedJSObject` is the same as `Foo.wrappedJSObject`, but for some, not. In fact, I run into this while developing that `MediaQueryList` and `EventTarget` hooks for rugk's project.

> but only as long as it doesn't cause security issues.

Don't worry about that. `Function.prototype.call.call(originalGetter, this);` (`_call.call(originalGetter, this)` is OK, but `originalGetter.call(this)` is not.

**Comment by tartpvule** [comment]
`Do you have any idea what might be causing that?`

Probably https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_scripts
Quote:
````
Note that in Firefox, content scripts won't be injected into empty iframes at "document_start", even if you specify that value in run_at .
````

Working around this is a herculean task (such as hooking `onload` events, `contentWindow` property, and others). Not worth the effort in my opinion. "Lies beget lies", you have to keep lying to cover the lies you already told.

**Comment by airtower-luna** [comment]
Thank you for that bug link, though it doesn't make me happy either! :disappointed_relieved: That means I'll have to adjust the test so it doesn't expect `document.referrer` manipulation on the iframe to work reliably. At least it still works much better than before on the main page.

This quote from the bug discussion unfortunately also means we mustn't assume that we get an unmodified getter after unwrapping `Document.prototype`:

> It's possible for same-origin pages to execute scripts in about:blank pages before extension content scripts, but that also goes for content scripts in any other pages.

File: content.js (hunk 9 above)
Comment (airtower-luna) [reply content.js:30]: I've been reading a lot more today and noticed this in the [Xray vision page](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/Xray_vision):

> It's worth emphasizing that even if content tricks chrome into running some unexpected code, that code does not run with chrome privileges.

So now I'm confident we're good. :slightly_smiling_face: I'll look into the test issue with iframes, if you could update the HTMLDocument to remove the prototype check we should be good to merge.

**Comment by tartpvule** [comment]
> we mustn't assume that we get an unmodified getter after unwrapping `Document.prototype`

Based on https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_scripts mentioning only "empty iframes" and my own (private) extensions in the past year that do attempt verify the DOM at the start of content scripts. I'm inclined to believe that we will truly get `document_start` as it should be in non-`about:blank` cases, if only because there is nothing we can do if that assumption is broken.

Specifically, if a hostile script can modify `get referrer` on `Document.prototype` before we get to it, then it surely can store the original function somewhere, or make that property non-configurable. We are too late and nothing we do will matter in this case.

In addition, the way we are using the return value of the possibly adulterated `get referrer` cannot possibly result in a confused deputy problem and cannot mutate any internal states of the content script. So we are clear on the security front, but detection is certainly possible, and we cannot prevent detection anyway if only due to timing side channels. So, preventing detection is, for me, best effort only and only if it does not unduly impact performance.

All in all, I think it is useless to dwell on the possibility that some code could run before the content script, as, due to JavaScript's immense malleability, there is nothing we can do. It is better to solve the solvable cases.

Footnote: There *IS* technically a way to solve the "blank iframe" problem: we hook *every* possible way hostile scripts can get a reference to other window/document. Needless to say, this is a *massive* undertaking and will touch at least 2 other parts of the DOM (`EventTarget`, `HTMLIFrameElement`) and possibly many others as a cascading effect of "covering lies".
But there will be nothing we can do if we cannot run code before any hostile code at top level, too.

Thinking again, it's probably better to have `String()` to coerce the return value to string type as a precaution against rules probing and prevent exceptions in *some* places, but this is ultimately useless if someone goes to the trouble of using this race window for detection: one simply has to compare (`===`) the hooked and the non-hooked `get referrer` and the game is up.

**Comment by tartpvule** [comment]
BTW This bypass also works:
```
var win = window.open('about:blank');
var real = Reflect.getOwnPropertyDescriptor(win.Document.prototype, 'referrer').get.call(document);
```
So... not just iframes, but also other ways of spawning new `window` with `about:blank`, at least.

File: content.js (hunk 9 above)
Comment (tartpvule) [reply content.js:30]: Do you mean this?
```
_toString.call(Reflect.getPrototypeOf(this)) !== "[object HTMLDocument]"
```

Comment (airtower-luna) [reply content.js:30]: The former. So the check should be just:
```js
if (Object.prototype.toString.call(this) !== "[object HTMLDocument]")
```

Because checking both is unnecessary, as you confirmed here: https://github.com/airtower-luna/referer-mod/pull/20#discussion_r612086397 :slightly_smiling_face:

Comment (tartpvule) [reply content.js:30]: ??? What about `HTMLDocument.prototype`? Ignore that case?
```
Reflect.getOwnPropertyDescriptor(win.Document.prototype, 'referrer').get.call(HTMLDocument.prototype);
```
`_toString.call(Reflect.getPrototypeOf(this)) !== "[object HTMLDocument]"` is specifically to check for this.

Comment (airtower-luna) [reply content.js:30]:
```js
Object.prototype.toString.call(HTMLDocument.prototype)
```
also returns `"[object HTMLDocument]"`, so if `this` is the prototype (or rather its wrapper) the simple check still works. Or do you have some other case in mind?

Comment (tartpvule) [reply content.js:30]: Those checks are to weed out everything that is not a document, or, in Mozilla terms, does not implement interface `Document`, which neither `HTMLDocument.prototype` nor `Document.prototype` are.
In normal Firefox, the only objects that implement this interface are `document` objects, whose prototypes are, confusingly, [`HTMLDocument`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDocument) whose prototypes are [`Document`](https://developer.mozilla.org/en-US/docs/Web/API/Document).
The only legitimate object (hance visible through X-ray vision) that `Object.prototype.toString.call(HTMLDocument.prototype) === "[object HTMLDocument]"` but does not internally implement the `Document` interface is `HTMLDocument.prototype`. So I check for it.

So, for `var x = HTMLDocument.prototype`:
```
Object.prototype.toString.call(x) === "[object HTMLDocument]"
Object.prototype.toString.call(Reflect.getPrototypeOf(this)) === "[object Document]"
```
But for `var x = document` or `var x = someIframe.contentWindow.document`:
```
Object.prototype.toString.call(x) === "[object HTMLDocument]"
Object.prototype.toString.call(Reflect.getPrototypeOf(this)) === "[object HTMLDocument]"
```

Comment (tartpvule) [reply content.js:30]: Addendum:
Forgot to mention the case where someone does `var foo = new Document();`
I think we should also passthrough `foo.referrer` in that case.

See: https://developer.mozilla.org/en-US/docs/Web/API/Document/Document

**Comment by tartpvule** [comment]
Is there something wrong? It fails at "Build testenv container images" and did not proceed to the tests themselves at all?

**Comment by airtower-luna** [comment]
It says "received unexpected HTTP status: 503 Service Unavailable" while pulling the `golang` base image, so I guess something's wrong with the registry. I'll re-run the workflow in a bit, hopefully they get it fixed meanwhile. :wink:

File: content.js (hunk 9 above)
Comment (airtower-luna) [reply content.js:30]: Okay, now I see your point, thanks. The `foo = new Document()` case should get caught by the existing checks, so `foo.referrer` will be unmodified. :slightly_smiling_face:

**Comment by tartpvule** [comment]
All green :smile:

**Comment by airtower-luna** [comment]
I've been experimenting with adding iframes through Javascript in the tests, based on the `window.open('about:blank');` example you posted. And unfortunately the bypass works not only on `about:blank` but also with `/` or a static HTML page I added to the test server. :confused: 

This reliably returns the real referrer:

```js
      function iframeRef()
      {
	  let frame = document.createElement("iframe");
	  frame.setAttribute("src", "/static/empty.html");
	  document.body.appendChild(frame);
	  let real = Reflect.getOwnPropertyDescriptor(frame.contentWindow.Document.prototype, 'referrer').get.call(document);
	  return real;
      }
      document.getElementById("referrer-iframe").innerText = iframeRef();
```

Almost like the content script doesn't get injected at all, or not until after that function is done.

**Comment by tartpvule** [comment]
Well... this gets complicated fast!

Short of that workaround I mentioned in "Footnote" above, I don't think we can do anything else.
This is hostile code accessing and mutating the environment before `document-start` (`readyState === "loading"`).

Try this: (tested on https://www.example.com/)
```
var frame = document.createElement("iframe");
frame.setAttribute('src', '/')
document.body.appendChild(frame);
function trap() {
 alert('BOOM');
}
Reflect.defineProperty(frame.contentWindow.Document.prototype, 'referrer', {
 configurable: true,
 enumerable: true,
 get: trap
});
setTimeout(function() {
 var temp = Reflect.getOwnPropertyDescriptor(frame.contentWindow.Document.prototype, 'referrer').get;
 temp.call(document);
}, 1000);
```

If you modify the hook in content.js to make it obvious that it is triggered (e.g. `alert("Triggered")` at the start of the hook), you will see that the `trap` function is indeed called from the hook.
Nothing we can do about this, and this impacts *EVERY* security WebExtensions that tries hooking in any way. We are not unique here.

I might research that "covering lies" I mentioned in the Footnote in https://github.com/airtower-luna/referer-mod/pull/20#issuecomment-820534524 later. But should that bear fruits, it will be in another PR.

To summarize what is happening:
1. `document.createElement("iframe")` -> `document.body.appendChild(frame)`
2. A `Window` goes live and can be accessed immediately with `frame.contentWindow`. AND `readyState === "uninitialized"` 😱
3. Content scripts are queued for execution at `readyState === "loading"` (See https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_scripts)

Note that the "uninitialized" readyState is **NOT** in the [WHATWG HTML spec](https://html.spec.whatwg.org/multipage/dom.html#documentreadystate) at all and is not even documented by [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Document/readyState). But see [w3schools.com](https://www.w3schools.com/jsref/prop_doc_readystate.asp) 😮.

WOW! We are in undocumented land. 😱

**Comment by airtower-luna** [comment]
Hi @tartpvule, sorry I haven't been responding in so long! Sad life events got in the way.

As you probably saw already I've pushed an update to the tests, unless you have any objections I'm going to merge this soon. Leaving the iframe manipulation as best-effort seems like the best available option to me, if there are new options I'd be happy to see. Thank you very much for all your work on this! :slightly_smiling_face:

**Review by airtower-luna — APPROVED** [review]

**Comment by tartpvule** [comment]
To be frank, I am not entirely satisfied with this, given that all a site has to do to completely bypass any security is to do https://github.com/airtower-luna/referer-mod/pull/20#issuecomment-821014848.
Maybe that is the perfectionist in me, but I think this is equivalent to leaving referrer exposed, which negates any security improvements.

Objectively speaking, this PR is an improvement and I don't have any objetions if you merge it. But I am not satisfied with "leaving the job half done", so expect another PR on this soon (TM). :smiley:

I am developing an experimental solution (https://github.com/airtower-luna/referer-mod/pull/20#issuecomment-820534524). It is... "elaborate" to say the least and probably involves some... "insanity".

So... merge, I think. But expect at least another PR to "finish the job". :smiley:

Nice to have worked with you!

**Comment by tartpvule** [comment]
Continue at https://github.com/airtower-luna/referer-mod/issues/17#issuecomment-843507333
